peer/x/peerheap/list.go

```diff
@@ -0,0 +1,242 @@
+// callback for when the request is finished, or an error if it fails.
+// The choose method takes a context that must have a deadline.
+// Choose resepects this deadline, waiting for an available peer until the
+// deadline.
+// The peer heap does not use the given *transport.Request and can safely
+// receive nil.
+func (pl *List) Choose(ctx context.Context, _ *transport.Request) (peer.Peer, func(error), error) {
+	if !pl.IsRunning() {
+		return nil, nil, peer.ErrPeerListNotStarted("PeerHeap")
+	}
+
+	for {
+		if ps, ok := pl.get(); ok {
+			pl.notifyPeerAvailable()
+			ps.peer.StartRequest()
+			return ps.peer, ps.boundFinish, nil
+		}
+
+		if err := pl.waitForPeerAvailableEvent(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+}
+
+func (pl *List) get() (*peerScore, bool) {
+	pl.mu.Lock()
+	defer pl.mu.Unlock()
+
+	ps, ok := pl.byScore.popPeer()
+	if !ok {
+		return nil, false
+	}
+
+	// Note: We push the peer back to reset the "next" counter.
+	// This gives us round-robin behavior.
+	pl.byScore.pushPeer(ps)
+
+	return ps, ps.status.ConnectionStatus == peer.Available
+}
+
+// waitForPeerAvailableEvent waits until a peer is added to the peer list or the
+// given context finishes.
+// Must NOT be run in a mutex.Lock()
+func (pl *List) waitForPeerAvailableEvent(ctx context.Context) error {
+	if _, ok := ctx.Deadline(); !ok {
+		return peer.ErrChooseContextHasNoDeadline("PeerHeap")
+	}
+
+	select {
+	case <-pl.peerAvailableEvent:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// notifyPeerAvailable writes to a channel indicating that a Peer is currently
+// available for requests
+func (pl *List) notifyPeerAvailable() {
+	select {
+	case pl.peerAvailableEvent <- struct{}{}:
+	default:
+	}
+}
+
+// NotifyStatusChanged receives notifications when a peer becomes available,
+// connected, unavailable, or when its pending request count changes.
+// This method satisfies peer.Subscriber and is only used for tests, since
+// the peer heap has a subscriber for each invividual peer.
+func (pl *List) NotifyStatusChanged(pid peer.Identifier) {
+	pl.mu.Lock()
+	ps := pl.byIdentifier[pid.Identifier()]
+	pl.mu.Unlock()
+	ps.NotifyStatusChanged(pid)
+}
+
+func (pl *List) notifyStatusChanged(ps *peerScore) {
+	pl.mu.Lock()
+	pl.rescorePeer(ps)
+	pl.mu.Unlock()
+
+	if ps.peer.Status().ConnectionStatus == peer.Available {
+		pl.notifyPeerAvailable()
+	}
+}
+
+func (pl *List) internalNotifyStatusChanged(ps *peerScore) {
+	pl.rescorePeer(ps)
+
+	if ps.peer.Status().ConnectionStatus == peer.Available {
+		pl.notifyPeerAvailable()
+	}
+}
+
+func (pl *List) rescorePeer(ps *peerScore) {
+	p := ps.peer
+	ps.status = p.Status()
+	ps.score = scorePeer(p)
+	pl.byScore.update(ps.idx)
+}
+
+func scorePeer(p peer.Peer) int64 {
+	status := p.Status()
+	score := int64(status.PendingRequestCount)
+	if status.ConnectionStatus != peer.Available {
+		score += int64(unavailablePenalty)
```

Comment:
@prashantv you’ll note that I simplified here. The TChannel peer currently begins life "Available", regardless of connection count, but the intention is to give the peer list responsibility for attempting to reconnect or releasing a peer in favor of another. This is an area I expect to revisit.